File: Samples/MicrosoftGraph.iOS.Objective-C.SendMail/MSGraph.MailClient/MailClient.m
//
//  MailClient.m
//  MSGraph.MailClient
//
//  Created by Simon Jaeger (@simonjaegr) on 08/03/16.
//  Copyright © 2016 Office 365 Development Patterns and Practises.
//

#import "MailClient.h"

@implementation MailClient
NSString *const GraphServiceUrl = @"https://graph.microsoft.com/";
NSString *const GraphServiceVersion = @"v1.0";

// Initialize ADAL (Active Directory Authentication Library) and the Microsoft Graph SDK.
- (id)init {
    self = [super init];
    if (self) {
        self.resolver = [[ADALDependencyResolver alloc] initWithPlist];
        self.graphClient = [[MSGraphServiceClient alloc] initWithUrl:[self getGraphServiceUrl]
                                                  dependencyResolver:self.resolver];
        self.ready = NO;
    }
    return self;
}

// Get the complete Microsoft Graph URL (including the version).
- (NSString *)getGraphServiceUrl {
    return [NSString stringWithFormat:@"%@%@", GraphServiceUrl, GraphServiceVersion];
}

// Let a user sign in, before we can call the Microsoft Graph.
- (void)signInWithCallback:(void (^)(BOOL))callback {
    [self.resolver interactiveLogonWithCallback:^(ADAuthenticationResult *result) {
        self.ready = result.status == AD_SUCCEEDED;
        
        // Call the callback so that we may continue.
        callback(self.ready);
    }];
}

// Get the current signed in user.
- (void)getUserWithCallback:(void (^)(MSGraphServiceUser *))callback {
    // Make sure that the client is ready.
    if (!self.ready || self.graphClient == nil) {
        NSLog(@"The mail client has not been initialized properly");
        
        // Nothing to return.
        callback(nil);
    }
    else {
        // Get the user.
        [self.graphClient.me readWithCallback:^(MSGraphServiceUser *user, MSOrcError *error) {
            // Return the user.
            callback(user);
        }];
    }
}

// Create a recipient object using an email address.
- (MSGraphServiceRecipient *)createRecipientWithAddress:(NSString *)address {
    // Create the email address object.
    MSGraphServiceEmailAddress *emailAddress = [MSGraphServiceEmailAddress new];
    emailAddress.address = address;
    
    // Create and return the recipient object.
    MSGraphServiceRecipient *recipient = [MSGraphServiceRecipient new];
    recipient.emailAddress = emailAddress;
    return recipient;
}

// Create a body object using HTML content.
-(MSGraphServiceItemBody *)createBodyWithContent:(NSString *)content {
    MSGraphServiceItemBody *body = [MSGraphServiceItemBody new];
    
    // Set the content type to HTML.
    body.contentType = MSGraphServiceBodyTypeHtml;
    
    // Set the HTML content.
    body.content = content;
    return body;
}

// Create a message object that can be used when sending a mail.
- (MSGraphServiceMessage *)createMessageWithSubject:(NSString *)subject
                                            content:(NSString *)content
                                         recipients:(NSArray *)recipients {
    MSGraphServiceMessage *message = [MSGraphServiceMessage new];
    
    // Set the message subject.
    message.subject = subject;
    
    // Set the importance of the message (Low, Normal, High).
    message.importance = MSGraphServiceImportanceNormal;
    
    
    // Convert the recipient strings to recipient objects used
    // with the Microsoft Graph.
    NSMutableArray *toRecipients = [NSMutableArray array];
    [recipients enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        [toRecipients addObject:[self createRecipientWithAddress:obj]];
    }];
    
    // Set the recipients of the message.
    message.toRecipients = toRecipients;
    
    // Create and set the body object.
    MSGraphServiceItemBody *body = [self createBodyWithContent:content];
    message.body = body;
    return message;
}

// Send a mail using a message object.
- (void)sendMailWithMessage:(MSGraphServiceMessage *)message
                   callback:(void (^)(BOOL))callback {
    // Make sure that the client is ready.
    if (!self.ready || self.graphClient == nil) {
        NSLog(@"The mail client has not been initialized properly");
        
        // Return failure.
        callback(NO);
    }
    else {
        // Send the mail.
        [self.graphClient.me.operations sendMailWithMessage:message
                                            saveToSentItems:YES
                                                   callback:^(int status, MSOrcError *error) {
            
            // Call the callback.
            callback(status == 0 && error == nil);
        }];
    }
}

// Send a mail to the current user.
- (void)sendMeWithCallback:(void (^)(BOOL))callback
                    logger:(void (^)(NSString *))logger {
    // Let a user sign in.
    logger(@"Signing in...");
    [self signInWithCallback:^(BOOL success) {
        if (!success) {
            callback(NO);
            return;
        }
        
        // Get the current user.
        logger(@"Getting user information...");
        
        [self getUserWithCallback:^(MSGraphServiceUser *user) {
            // Validate the user.
            if (user == nil || user.mail == nil) {
                callback(NO);
                return;
            }
            
            // Create the message content.
            NSMutableString *content = [NSMutableString string];
            [content appendString:@"<strong>Lorem ipsum dolor sit amet</strong>, consectetur adipiscing "];
            [content appendString:@"elit, sed do eiusmod tempor incididunt ut labore et dolore "];
            [content appendString:@"magna aliqua. Ut enim ad minim veniam, quis nostrud "];
            [content appendString:@"exercitation ullamco laboris nisi ut aliquip ex ea commodo "];
            [content appendString:@"consequat. Duis aute irure dolor in reprehenderit in voluptate "];
            [content appendString:@"velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint "];
            [content appendString:@"occaecat cupidatat non proident, sunt in culpa qui officia deserunt "];
            [content appendString:@"mollit anim id est laborum."];
            
            // Create the message.
            MSGraphServiceMessage *message = [self createMessageWithSubject:@"Hello #Office365Dev"
                                                                    content:content
                                                                 recipients:[NSArray arrayWithObject:user.mail]];
            
            // Send the message.
            logger(@"Sending mail...");
            [self sendMailWithMessage:message callback:callback];
        }];
    }];
}
@end

File: Samples/MicrosoftGraph.iOS.Swift.SendMail/MSGraph.MailClient/MailClient.swift
//
//  MailClient.swift
//  MSGraph.MailClient
//
//  Created by Simon Jaeger (@simonjaegr) on 07/03/16.
//  Copyright © 2016 Office 365 Development Patterns and Practises.
//

import Foundation

struct MailClientConstants {
    static let graphServiceUrl = "https://graph.microsoft.com/"
    static let graphServiceVersion = "v1.0"
}

class MailClient {
    var resolver: ADALDependencyResolver?
    var graphClient: MSGraphServiceClient?
    var ready = false
    
    // Initialize ADAL (Active Directory Authentication Library) and the Microsoft Graph SDK.
    init() {
        self.resolver = ADALDependencyResolver(plist: ())
        self.graphClient = MSGraphServiceClient(url: self.getGraphServiceUrl(), dependencyResolver: self.resolver)
        self.ready = false
    }
    
    // Get the complete Microsoft Graph URL (including the version).
    func getGraphServiceUrl() -> String {
        return MailClientConstants.graphServiceUrl + MailClientConstants.graphServiceVersion
    }
    
    // Let a user sign in, before we can call the Microsoft Graph.
    func signInWithCallback(callback: (Bool) -> Void) {
        self.resolver?.interactiveLogonWithCallback { (result: ADAuthenticationResult!) -> Void in
            self.ready = result.status == AD_SUCCEEDED
            
            // Call the callback so that we may continue.
            callback(self.ready)
        }
    }
    
    // Get the current signed in user.
    func getUserWithCallback(callback: (MSGraphServiceUser!) -> Void) {
        // Make sure that the client is ready.
        if (!self.ready || self.graphClient == nil) {
            print("The mail client has not been initialized properly")
            
            // Nothing to return.
            callback(nil)
        }
        else {
            // Get the user
            self.graphClient?.me.readWithCallback { (user: MSGraphServiceUser!, error: MSOrcError!) -> Void in
                // Return the user.
                callback(user)
            }
        }
    }
    
    // Create a recipient object using an email address.
    func createRecipientWithAddress(address: String) -> MSGraphServiceRecipient {
        // Create the email address object.
        let emailAddress = MSGraphServiceEmailAddress()
        emailAddress.address = address
        
        // Create and return the recipient object.
        let recipient = MSGraphServiceRecipient()
        recipient.emailAddress = emailAddress
        return recipient
    }
    
    // Create a body object using HTML content.
    func createBodyWithContent(content: String) -> MSGraphServiceItemBody {
        let body = MSGraphServiceItemBody()
        
        // Set the content type to HTML.
        body.contentType = .Html
        
        // Set the HTML content.
        body.content = content
        return body
    }
    
    // Create a message object that can be used when sending a mail.
    func createMessageWithSubject(subject: String, content: String, recipients: [String]) -> MSGraphServiceMessage {
        let message = MSGraphServiceMessage()
        
        // Set the message subject.
        message.subject = subject
        
        // Set the importance of the message (Low, Normal, High).
        message.importance = .Normal
        
        // Convert the recipient strings to recipient objects used
        // with the Microsoft Graph.
        let toRecipients = recipients.map({ (recipient: String) -> MSGraphServiceRecipient in return self.createRecipientWithAddress(recipient) })
        
        // Set the recipients of the message.
        message.toRecipients = NSMutableArray(array: toRecipients)
        
        // Create and set the body object.
        let body = self.createBodyWithContent(content)
        message.body = body
        return message
    }
    
    // Send a mail using a message object.
    func sendMailWithMessage(message: MSGraphServiceMessage, callback: (Bool) -> Void) {
        // Make sure that the client is ready.
        if (!self.ready || self.graphClient == nil) {
            print("The mail client has not been initialized properly")
            
            // Return failure.
            callback(false)
        }
        else {
            // Send the mail.
            self.graphClient?.me.operations.sendMailWithMessage(message, saveToSentItems: true, callback: { (status: Int32, error: MSOrcError!) -> Void in
                
                // Call the callback.
                callback(status == 0 && error == nil)
            })
        }
    }
    
    // Send a mail to the current user.
    func sendMeWithCallback(callback: (Bool) -> Void, logger: (String) -> Void) {
        // Let a user sign in.
        logger("Signing in...")
        self.signInWithCallback { (success: Bool) -> Void in
            if (!success) {
                callback(false)
                return
            }
            
            // Get the current user.
            logger("Getting user information...")
            self.getUserWithCallback { (user: MSGraphServiceUser!) -> Void in
                // Validate the user.
                if (user == nil || user.mail == nil) {
                    callback(false)
                    return
                }
                
                // Create the message.
                let message = self.createMessageWithSubject("Hello #Office365Dev", content:
                    "<strong>Lorem ipsum dolor sit amet</strong>, consectetur adipiscing " +
                        "elit, sed do eiusmod tempor incididunt ut labore et dolore " +
                        "magna aliqua. Ut enim ad minim veniam, quis nostrud " +
                        "exercitation ullamco laboris nisi ut aliquip ex ea commodo " +
                        "consequat. Duis aute irure dolor in reprehenderit in voluptate " +
                        "velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint " +
                        "occaecat cupidatat non proident, sunt in culpa qui officia deserunt " +
                    "mollit anim id est laborum.", recipients: [user.mail])
                
                // Send the message.
                logger("Sending mail...")
                self.sendMailWithMessage(message, callback: callback)
            }
        }
    }
}